### PR TITLE
DrivAerML: 4L/320d + gc=1.5 + WD=1e-2 (compound AirfRANS wins)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
On AirfRANS, grad-clip=1.5 alone gave a 26.4% improvement and WD=1e-2 showed strong regularization. Combining both on DrivAerML tests whether these improvements are additive or partially redundant. Both changes address different failure modes (gradient instability vs. weight norm growth), so they may compound. This is the compound test of the two strongest AirfRANS interventions applied to the 4L/320d DrivAerML configuration.

## Instructions
Run from `target/icml2026` with the following command (copy exactly):

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 320 \
  --model-heads 5 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --grad-clip 1.5 \
  --weight-decay 1e-2 \
  --wandb_group drivaerml-4L320d-compound
```

Changes from baseline: added `--grad-clip 1.5` and `--weight-decay 1e-2`. All other hyperparameters unchanged — this isolates the compound regularization effect.

Note: shouko is running WD=1e-2 alone (PR #2752). Your results alongside hers will tell us whether grad-clip adds value on top of WD, or whether WD alone is sufficient.

## Baseline
Current DrivAerML best (PR #2648, W&B run `qx7z7if3`):
- `val_primary/surface_rel_l2_pct`: **5.027%** (val) / 6.244% (test)
- Epoch: 257 of 180-min budget
- External competition target: **< 3.71%**

Baseline reproduce command:
```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

Beat `val_primary/surface_rel_l2_pct < 5.027%` to be a winner.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2761 because ADVISOR: Closing — double dead end. gc=1.5 is only viable at 3L/192d (5 confirmations of failure at 4L+), AND WD=1e-2 catastrophically diverges on DrivAerML. Baseline has also moved to 4L/512d..

**Comment by tcapelle** [comment]
STUDENT shinobu:

## Results

### Metrics

| Metric | Baseline (4L/320d, no gc/wd) | This run (gc=1.5 + WD=1e-2) | Delta |
|--------|-----|-----|-------|
| **val_primary/surface_rel_l2_pct** | 5.027% (ep 257) | **8.88%** (ep 79, pre-divergence) | **+3.85 pp (worse)** |
| **test_primary/surface_rel_l2_pct** | 6.244% | **80.69%** (post-divergence) | **+74.45 pp** |

### NEGATIVE RESULT — Model diverged at epoch 88

The compound of grad-clip=1.5 + WD=1e-2 causes catastrophic gradient instability on DrivAerML around epoch 88. The model never recovers.

### Divergence timeline

| Epoch | val_primary | grad_clip_events | grad_norm_mean |
|-------|-------------|-----------------|----------------|
| 85 | 10.23% | 4 | 0.48 |
| 86 | 8.93% | 2 | 0.45 |
| 87 | 11.16% | 3 | 0.51 |
| **88** | **31.69%** | **103** | **4.24** |
| 89 | 35.70% | 358 | 11.2 |
| 90 | 17.73% | 232 | 9.3 |
| 93 | 59.51% | 361 | 84.0 |
| 94 | 62.05% | 383 | 158.0 |
| 95+ | ~79-80% | 394 | Infinity |

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 320 \
  --model-heads 5 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --grad-clip 1.5 \
  --weight-decay 1e-2 \
  --agent shinobu \
  --wandb-name "shinobu/drivaerml-4L320d-gc15-wd1e2" \
  --wandb-group "drivaerml-4L320d-compound"
```

### Details

- **W&B run:** `ykcp9l6c` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ykcp9l6c))
- **Peak memory:** ~6.9 GB GPU VRAM (GPU 0, NVIDIA RTX PRO 6000 Blackwell)
- **Epochs completed:** 267 (hit 180-min timeout)
- **Time per epoch:** ~40 seconds
- **Best val (pre-divergence):** 8.88% at epoch 79

### What happened

**The hypothesis is rejected.** The compound of grad-clip=1.5 and WD=1e-2 is catastrophically unstable on DrivAerML. While both interventions worked well individually on AirfRANS, their combination causes gradient explosion on DrivAerML at epoch 88.

Key observations:

1. **Abrupt divergence.** Training was progressing normally (val 8.88-11%) through epoch 87 with only 2-4 clip events per epoch. At epoch 88, clip events jumped to 103 and grad norms exploded. By epoch 95, all 394 training steps were being clipped every epoch and grad norms reached Infinity.

2. **Never recovers.** Despite continued training for 179 more epochs after divergence, the model never recovered. Val stabilized at ~79-80% — essentially random predictions.

3. **WD=1e-2 likely the culprit.** The default WD is 1e-4, so 1e-2 is a 100x increase. On DrivAerML (which has much higher-dimensional inputs than AirfRANS with 50k surface points), aggressive weight decay may cause weight matrices to shrink too much, creating numerical instability when combined with gradient clipping that prevents corrective large updates.

4. **Even pre-divergence, not competitive.** Best val was 8.88% at epoch 79, already 3.85 pp worse than the 5.027% baseline at the same epoch count.

### Suggested follow-ups

1. **Grad-clip=1.5 alone (no WD change)** — isolate the grad-clip effect. taki (PR #2xxx) may already be testing this.
2. **WD=1e-3 (10x instead of 100x)** — a milder WD increase might regularize without causing instability.
3. **WD=1e-2 alone (no grad-clip)** — shouko is testing this (PR #2752). Compare to see if the instability comes from WD alone or the compound.
